src/promptflow-devkit/promptflow/_sdk/_utilities/serve_utils.py

```diff
@@ -22,6 +24,154 @@
 logger = logging.getLogger(__name__)
 
 
+class ServeAppHelper(abc.ABC):
+    """The abstract class for serve app helper.
+
+    help to start and terminate the serve app.
+    """
+
+    def __init__(
+        self, *, flow_file_name: str, flow_dir: Path, init: Dict[str, Any], port: int, host: str = "localhost", **kwargs
+    ):
+        self._flow_file_name = flow_file_name
+        self._flow_dir = flow_dir
+        self._init = init or {}
+        self._port = port
+        self._host = host
+
+    @abc.abstractmethod
+    def start_in_main(self, skip_open_browser: bool = False):
+        """Start the serve app in main thread."""
+        pass
+
+    @abc.abstractmethod
+    def start(self):
+        """Start the serve app in a subprocess."""
+        pass
+
+    @abc.abstractmethod
+    def terminate(self):
+        """Terminate the serve app in subprocess."""
+        pass
+
+
+class PythonServeAppHelper(ServeAppHelper):
+    def __init__(
+        self, *, flow_file_name: str, flow_dir: Path, init: Dict[str, Any], port: int, host: str = "localhost", **kwargs
+    ):
+        self._static_folder: Optional[str] = kwargs.get("static_folder", None)
+        self._config = kwargs.get("config", {}) or {}
+        self._environment_variables = kwargs.get("environment_variables", {}) or {}
+        self._engine = kwargs.get("engine", "flask")
+
+        super().__init__(flow_file_name=flow_file_name, flow_dir=flow_dir, init=init, port=port, host=host, **kwargs)
+
+        self._process: Optional[multiprocessing.Process] = None
+
+    def _run(self, skip_open_browser: bool = False):
+        from promptflow._sdk._configuration import Configuration
+        from promptflow.core._serving.app import create_app
+
+        pf_config = Configuration(overrides=self._config)
+        logger.info(f"Promptflow config: {pf_config}")
+        connection_provider = pf_config.get_connection_provider()
+        os.environ["PROMPTFLOW_PROJECT_PATH"] = self._flow_dir.absolute().as_posix()
+        logger.info(f"Change working directory to model dir {self._flow_dir}")
+        os.chdir(self._flow_dir)
+        app = create_app(
+            static_folder=self._static_folder,
+            environment_variables=self._environment_variables,
+            connection_provider=connection_provider,
+            init=self._init,
+        )
+        if not skip_open_browser:
+            target = f"http://{self._host}:{self._port}"
+            logger.info(f"Opening browser {target}...")
+            webbrowser.open(target)
+
+        if self._engine == "flask":
```

Comment:
could this logic be in promptflow.core

Reply:
+ @wxpjimmy for this. I just move it from a simple function to a class and got no idea about why we should put this logic here

Reply:
let me check if it's possible to avoid moving the code